lib/types.nix

```
@@ -1233,13 +1252,25 @@ let
                 docsEval = base.extendModules { modules = [ noCheckForDocsModule ]; };
               in
               docsEval._module.freeformType.description or name;
-          check = x: isAttrs x || isFunction x || path.check x;
-          merge =
-            loc: defs:
-            (base.extendModules {
-              modules = [ { _module.args.name = last loc; } ] ++ allModules defs;
-              prefix = loc;
-            }).config;
+          inherit check;
+          merge = {
+            __functor =
+              self: loc: defs:
+              (self.v2 { inherit loc defs; }).value;
+            v2 =
+              { loc, defs }:
+              let
+                configuration = base.extendModules {
+                  modules = [ { _module.args.name = last loc; } ] ++ allModules defs;
+                  prefix = loc;
+                };
+              in
+              {
+                headError = checkDefsForError check loc defs;
+                value = configuration.config;
+                valueMeta = configuration;
```

This will be invaluable :rocket:

---

This line is the core motivation!, along with the analogous one for attribute sets.

------

`configuration` has a `_type` that we shouldn't forward here.
I think it would also be wise to make the forwarding of each individual attribute a conscious decision.
`type` is also confusing in this context, because it includes the definitions as part of the "static module", so let's leave it out for now.

```suggestion
                valueMeta = {
                  inherit (configuration) config options extendModules;
                };
```

Not sure about `class` because it's static, and available in `type.functor.payload.class`.
Should we add `type` instead? Seems generally useful.